File: mobile/src/main/java/org/openhab/habdroid/ui/CloudNotificationListFragment.java
/*
 * Copyright (c) 2010-2016, openHAB.org and others.
 *
 *   All rights reserved. This program and the accompanying materials
 *   are made available under the terms of the Eclipse Public License v1.0
 *   which accompanies this distribution, and is available at
 *   http://www.eclipse.org/legal/epl-v10.html
 */

package org.openhab.habdroid.ui;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import okhttp3.Call;
import okhttp3.Headers;
import okhttp3.Request;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.openhab.habdroid.R;
import org.openhab.habdroid.core.connection.Connection;
import org.openhab.habdroid.core.connection.ConnectionFactory;
import org.openhab.habdroid.model.CloudNotification;
import org.openhab.habdroid.ui.widget.DividerItemDecoration;
import org.openhab.habdroid.util.AsyncHttpClient;
import org.openhab.habdroid.util.Util;

import java.util.ArrayList;
import java.util.Locale;

public class CloudNotificationListFragment extends Fragment implements
        View.OnClickListener, SwipeRefreshLayout.OnRefreshListener {

    private static final String TAG = CloudNotificationListFragment.class.getSimpleName();

    private static final int PAGE_SIZE = 20;

    private MainActivity mActivity;
    // keeps track of current request to cancel it in onPause
    private Call mRequestHandle;

    private CloudNotificationAdapter mNotificationAdapter;
    private String mInitiallyHighlightedId;

    private RecyclerView mRecyclerView;
    private LinearLayoutManager mLayoutManager;
    private SwipeRefreshLayout mSwipeLayout;
    private View mEmptyView;
    private ImageView mEmptyWatermark;
    private TextView mEmptyMessage;
    private View mRetryButton;
    private int mLoadOffset;

    public static CloudNotificationListFragment newInstance(@Nullable String highlightedId) {
        CloudNotificationListFragment f = new CloudNotificationListFragment();
        Bundle args = new Bundle();
        args.putString("highlightedId", highlightedId);
        f.setArguments(args);
        return f;
    }

    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public CloudNotificationListFragment() {
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        Log.i(TAG, "onCreateView");
        View view = inflater.inflate(R.layout.fragment_notificationlist, container, false);
        mSwipeLayout = view.findViewById(R.id.swipe_container);
        mSwipeLayout.setOnRefreshListener(this);
        Util.applySwipeLayoutColors(mSwipeLayout, R.attr.colorPrimary, R.attr.colorAccent);

        mRecyclerView = view.findViewById(android.R.id.list);
        mEmptyView = view.findViewById(android.R.id.empty);
        mEmptyMessage = view.findViewById(R.id.empty_message);
        mEmptyWatermark = view.findViewById(R.id.watermark);
        mRetryButton = view.findViewById(R.id.retry_button);
        mRetryButton.setOnClickListener(this);
        return view;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mActivity = (MainActivity) getActivity();
        mNotificationAdapter = new CloudNotificationAdapter(mActivity,
                () -> loadNotifications(false));
        mLayoutManager = new LinearLayoutManager(mActivity);

        mRecyclerView.setLayoutManager(mLayoutManager);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(mActivity));
        mRecyclerView.setAdapter(mNotificationAdapter);
        Log.d(TAG, "onActivityCreated()");
    }

    @Override
    public void onResume() {
        super.onResume();
        Log.d(TAG, "onResume()");
        loadNotifications(true);
    }

    @Override
    public void onPause() {
        super.onPause();
        Log.d(TAG, "onPause()");
        // Cancel request for notifications if there was any
        if (mRequestHandle != null) {
            mRequestHandle.cancel();
        }
    }

    @Override
    public void onRefresh() {
        Log.d(TAG, "onRefresh()");
        loadNotifications(true);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        Log.d(TAG, "onDetach()");
        mActivity = null;
    }

    @Override
    public void onClick(View view) {
        if (view == mRetryButton) {
            loadNotifications(true);
        }
    }

    private void loadNotifications(boolean clearExisting) {
        Connection conn = ConnectionFactory.getConnection(Connection.TYPE_CLOUD);
        if (conn == null) {
            updateViewVisibility(false, true);
            return;
        }
        if (clearExisting) {
            mNotificationAdapter.clear();
            mLoadOffset = 0;
            updateViewVisibility(true, false);
        }

        // If we're passed an ID to be highlighted initially, we'd theoretically need to load all
        // items instead of loading page-wise. As the initial highlight is only needed for
        // notifications and a new notification is very likely to be contained in the first page,
        // we skip that additional effort.
        final String url = String.format(Locale.US, "api/v1/notifications?limit=%d&skip=%d",
                PAGE_SIZE, mLoadOffset);
        final AsyncHttpClient client = conn.getAsyncHttpClient();
        mRequestHandle = client.get(url, new AsyncHttpClient.StringResponseHandler() {
            @Override
            public void onSuccess(String responseBody, Headers headers) {
                try {
                    ArrayList<CloudNotification> items = new ArrayList<>();
                    JSONArray jsonArray = new JSONArray(responseBody);
                    for (int i = 0; i < jsonArray.length(); i++) {
                        JSONObject sitemapJson = jsonArray.getJSONObject(i);
                        items.add(CloudNotification.fromJson(sitemapJson));
                    }
                    Log.d(TAG, "Notifications request success, got " + items.size() + " items");
                    mLoadOffset += items.size();
                    mNotificationAdapter.addLoadedItems(items, items.size() == PAGE_SIZE);
                    handleInitialHighlight();
                    updateViewVisibility(false, false);
                } catch (JSONException e) {
                    Log.d(TAG, "Notification response could not be parsed", e);
                    updateViewVisibility(false, true);
                }
            }

            @Override
            public void onFailure(Request request, int statusCode, Throwable error) {
                updateViewVisibility(false, true);
                Log.e(TAG, "Notifications request failure", error);
            }
        });
    }

    private void handleInitialHighlight() {
        Bundle args = getArguments();
        String highlightedId = args.getString("highlightedId");
        if (TextUtils.isEmpty(highlightedId)) {
            return;
        }

        final int position = mNotificationAdapter.findPositionForId(highlightedId);
        if (position >= 0) {
            mLayoutManager.scrollToPositionWithOffset(position, 0);
            mRecyclerView.postDelayed(() -> mNotificationAdapter.highlightItem(position), 600);
        }

        // highlight only once
        args.remove("highlightedId");
    }

    private void updateViewVisibility(boolean loading, boolean loadError) {
        boolean showEmpty = !loading && (mNotificationAdapter.getItemCount() == 0 || loadError);
        mRecyclerView.setVisibility(showEmpty ? View.GONE : View.VISIBLE);
        mEmptyView.setVisibility(showEmpty ? View.VISIBLE : View.GONE);
        mSwipeLayout.setRefreshing(loading);
        mEmptyMessage.setText(
                loadError ? R.string.notification_list_error : R.string.notification_list_empty);
        mEmptyWatermark.setImageResource(
                loadError ? R.drawable.ic_connection_error : R.drawable.ic_no_notifications);
        mRetryButton.setVisibility(loadError ? View.VISIBLE : View.GONE);
    }
}

File: mobile/src/main/java/org/openhab/habdroid/ui/CloudNotificationListFragment.kt
/*
 * Copyright (c) 2010-2024 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.ui

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.os.bundleOf
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.faltenreich.skeletonlayout.SkeletonLayout
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONException
import org.openhab.habdroid.R
import org.openhab.habdroid.core.connection.ConnectionFactory
import org.openhab.habdroid.model.CloudMessage
import org.openhab.habdroid.model.ServerConfiguration
import org.openhab.habdroid.model.toCloudMessage
import org.openhab.habdroid.util.HttpClient
import org.openhab.habdroid.util.getActiveServerId
import org.openhab.habdroid.util.getConfiguredServerIds
import org.openhab.habdroid.util.getPrefs
import org.openhab.habdroid.util.getPrimaryServerId
import org.openhab.habdroid.util.getSecretPrefs
import org.openhab.habdroid.util.map

/**
 * Mandatory empty constructor for the fragment manager to instantiate the
 * fragment (e.g. upon screen orientation changes).
 */
class CloudNotificationListFragment :
    Fragment(),
    View.OnClickListener,
    SwipeRefreshLayout.OnRefreshListener {
    lateinit var recyclerView: RecyclerView
    private lateinit var skeleton: SkeletonLayout
    private lateinit var swipeLayout: SwipeRefreshLayout
    private lateinit var retryButton: Button
    private lateinit var emptyView: View
    private lateinit var emptyWatermark: ImageView
    private lateinit var emptyMessage: TextView

    // keeps track of current request to cancel it in onPause
    private var requestJob: Job? = null
    private lateinit var adapter: CloudNotificationAdapter
    private lateinit var layoutManager: LinearLayoutManager
    private var loadOffset: Int = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        Log.i(TAG, "onCreateView")
        val view = inflater.inflate(R.layout.fragment_notificationlist, container, false)

        recyclerView = view.findViewById(android.R.id.list)
        emptyView = view.findViewById(android.R.id.empty)
        emptyMessage = view.findViewById(R.id.empty_message)
        emptyWatermark = view.findViewById(R.id.watermark)
        skeleton = view.findViewById(R.id.skeletonLayout)
        view.findViewById<LinearLayout>(R.id.skeletonList).apply {
            repeat(10) {
                addView(inflater.inflate(R.layout.notificationlist_item, null))
            }
        }
        skeleton.showSkeleton()

        swipeLayout = view.findViewById(R.id.swipe_container)
        swipeLayout.setOnRefreshListener(this)
        swipeLayout.applyColors()

        retryButton = view.findViewById(R.id.retry_button)
        retryButton.setOnClickListener(this)

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = CloudNotificationAdapter(view.context) { loadNotifications(false) }
        layoutManager = LinearLayoutManager(view.context)

        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = adapter
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume()")
        loadNotifications(true)
    }

    override fun onPause() {
        super.onPause()
        Log.d(TAG, "onPause()")
        // Cancel request for notifications if there was any
        requestJob?.cancel()
    }

    override fun onRefresh() {
        Log.d(TAG, "onRefresh()")
        swipeLayout.isRefreshing = false
        loadNotifications(true)
    }

    override fun onClick(view: View) {
        if (view === retryButton) {
            loadNotifications(true)
        }
    }

    private fun loadNotifications(clearExisting: Boolean) {
        val activity = activity as AbstractBaseActivity? ?: return
        val conn = if (usePrimaryServer()) {
            ConnectionFactory.primaryCloudConnection?.connection
        } else {
            ConnectionFactory.activeCloudConnection?.connection
        }
        if (conn == null) {
            updateViewVisibility(loading = false, loadError = true)
            return
        }
        if (clearExisting) {
            adapter.clear()
            loadOffset = 0
            updateViewVisibility(loading = true, loadError = false)
        }

        // If we're passed an ID to be highlighted initially, we'd theoretically need to load all
        // items instead of loading page-wise. As the initial highlight is only needed for
        // notifications and a new notification is very likely to be contained in the first page,
        // we skip that additional effort.
        val url = "api/v1/notifications?limit=$PAGE_SIZE&skip=$loadOffset"
        requestJob = activity.launch {
            try {
                val response = conn.httpClient.get(url).asText().response
                val allItems = JSONArray(response).map { obj -> obj.toCloudMessage() }
                val filteredItems = allItems.filterIsInstance<CloudMessage.CloudNotification>()
                Log.d(TAG, "Notifications request success, got ${allItems.size} items")
                loadOffset += allItems.size
                adapter.addLoadedItems(filteredItems, allItems.size == PAGE_SIZE)
                handleInitialHighlight()
                updateViewVisibility(loading = false, loadError = false)
            } catch (e: JSONException) {
                updateViewVisibility(loading = false, loadError = true)
                Log.d(TAG, "Notification response could not be parsed", e)
            } catch (e: HttpClient.HttpException) {
                updateViewVisibility(loading = false, loadError = true)
                Log.e(TAG, "Notifications request failure", e)
            }
        }
    }

    private fun handleInitialHighlight() {
        val highlightedId = arguments?.getString("highlightedId") ?: return
        val position = adapter.findPositionForId(highlightedId)
        if (position >= 0) {
            layoutManager.scrollToPositionWithOffset(position, 0)
            recyclerView.postDelayed({ adapter.highlightItem(position) }, 600)
        }

        // highlight only once
        arguments?.remove("highlightedId")
    }

    private fun updateViewVisibility(loading: Boolean, loadError: Boolean) {
        val showEmpty = !loading && (adapter.itemCount == 0 || loadError)
        recyclerView.isVisible = !showEmpty
        emptyView.isVisible = showEmpty
        skeleton.isVisible = loading
        emptyMessage.setText(
            if (loadError) R.string.notification_list_error else R.string.notification_list_empty
        )
        emptyWatermark.setImageResource(
            if (loadError) R.drawable.ic_connection_error else R.drawable.ic_no_notifications
        )
        retryButton.isVisible = loadError
    }

    private fun usePrimaryServer() = requireArguments().getBoolean("primary")

    fun getTitle(context: Context): String {
        val prefs = context.getPrefs()
        var title = context.getString(R.string.app_notifications)
        if (prefs.getConfiguredServerIds().size > 1) {
            val serverId = if (usePrimaryServer()) prefs.getPrimaryServerId() else prefs.getActiveServerId()
            val serverName = ServerConfiguration.load(prefs, context.getSecretPrefs(), serverId)?.name
            title = context.getString(R.string.ui_on_server, title, serverName)
        }

        return title
    }

    companion object {
        const val PAGE_SIZE = 50

        private val TAG = CloudNotificationListFragment::class.java.simpleName

        fun newInstance(highlightedId: String?, primaryServer: Boolean): CloudNotificationListFragment {
            val f = CloudNotificationListFragment()
            f.arguments = bundleOf("highlightedId" to highlightedId, "primary" to primaryServer)
            return f
        }
    }
}
